src/components/datepicker/datePicker.js

```diff
@@ -365,55 +535,108 @@
+
+      // TODO(jelbourn): Change this to classList.toggle when we stop using PhantomJS in unit tests
+      // because it doesn't conform to the DOMTokenList spec.
+      // See https://github.com/ariya/phantomjs/issues/12782.
+      if (!this.ngModelCtrl.$valid) {
+        this.inputContainer.classList.add(INVALID_CLASS);
       }
     } else {
-      // The date is seen as "not a valid date" if there is *something* set
-      // (i.e.., not null or undefined), but that something isn't a valid date.
-      this.ngModelCtrl.$setValidity('valid', date == null);
-    }
+      var beginDate = opt_begin_date || this.beginDate;
+      var endDate = opt_end_date || this.endDate;
+
+      // Clear any existing errors to get rid of anything that's no longer relevant.
+      this.clearErrorState();
+
+      if (this.dateUtil.isValidDate(beginDate) && this.dateUtil.isValidDate(endDate)) {
+        // Force all dates to midnight in order to ignore the time portion.
+        beginDate = this.dateUtil.createDateAtMidnight(beginDate);
+        endDate = this.dateUtil.createDateAtMidnight(endDate);
+
+        if (this.dateUtil.isValidDate(this.minDate)) {
+          var minDate = this.dateUtil.createDateAtMidnight(this.minDate);
+          this.setDateRangeValidity('mindate', beginDate >= minDate);
+        }
+
+        if (this.dateUtil.isValidDate(this.maxDate)) {
+          var maxDate = this.dateUtil.createDateAtMidnight(this.maxDate);
+          this.setDateRangeValidity('maxdate', endDate <= maxDate);
+        }
+
+        if (angular.isFunction(this.dateFilter)) {
+          if (this.dateFilter(beginDate) && this.dateFilter(endDate)) {
+            this.setDateRangeValidity('filtered', true);
+          } else {
+            this.setDateRangeValidity('filtered', false);
+          }
+        }
+      } else {
+        // The date range is seen as "not a valid date range" if there is *something* set
+        // (i.e.., not null or undefined), but that something isn't a valid date range.
+        this.setDateRangeValidity('valid', beginDate == null && endDate == null);
+      }
 
-    // TODO(jelbourn): Change this to classList.toggle when we stop using PhantomJS in unit tests
-    // because it doesn't conform to the DOMTokenList spec.
-    // See https://github.com/ariya/phantomjs/issues/12782.
-    if (!this.ngModelCtrl.$valid) {
-      this.inputContainer.classList.add(INVALID_CLASS);
+      if (!this.getDateRangeValidity().valid) {
+        this.inputContainer.classList.add(INVALID_CLASS);
+      }
     }
   };
 
   /** Clears any error flags set by `updateErrorState`. */
   DatePickerCtrl.prototype.clearErrorState = function() {
     this.inputContainer.classList.remove(INVALID_CLASS);
-    ['mindate', 'maxdate', 'filtered', 'valid'].forEach(function(field) {
-      this.ngModelCtrl.$setValidity(field, true);
-    }, this);
+    if (!this.isDateRangePicker) {
+      ['mindate', 'maxdate', 'filtered', 'valid'].forEach(function(field) {
```

Comment:
No reason to duplicate this array. And why check `isDateRange` vs. always clearing both?